### ISSUE TITLE
How to use Managed Identity authentication for AzureAD workload

### ISSUE THREAD

**Issue by Mohankade** [microsoft/Microsoft365DSC#3924]

**Comment by andikrueger** [comment]
Could you outline in which scenario you try to use managed identities? In a first place you need to have an environment, that supports managed identities (Azure Vm, Runbook, Function, Pipeline). After that you need to configure the roles of this identity: https://microsoft365dsc.com/concepts/personas/ 

Within M365DSC you need to add either the -ManagedIdentity switch or the parameter in your configuration to tell M365DSC to use MI.

**Comment by Mohankade** [comment]
Thank You for your response! We are planning to use Azure Automation Account with hybrid worker to run DSC configurations. 

Also, I have another questions on how to use managed identity across tenants for M365DSC?

**Comment by andikrueger** [comment]
As of now, managed identity cannot be used across tenants. There are Blog post explaining an approach on how to use as federated credentials. If this would work in M365DSC I cannot tell. presumably it would fail due to the missing tenant information about the other tenant.

**Comment by raandree** [comment]
When using the [DscWorkshop](https://github.com/dsccommunity/DscWorkshop) template for Micosoft365DSC you can control as many tenants as you want: [Microsoft365DscWorkshop](https://github.com/raandree/Microsoft365DscWorkshop).

It works like this: The configuration for all tenants is in one Yaml hierarchy to make sure they all get the same baseline and are in sync. For each tenant you need to have a self-hosted Azure DevOps agent (VM) with a managed identity assigned.

Through the build pipeline's parameter `buildEnvironments` you can control for which environment you want to build and what the dependencies are. In the proof of concept the default is this:

```yaml
- Name: Dev
- Name: Test
  DependsOn: Dev
- Name: Prod
  DependsOn: Test
```

In Azure Pipelines, each job is forwarded to the correct build agent using [demands](https://learn.microsoft.com/en-us/azure/devops/pipelines/yaml-schema/pool-demands?view=azure-pipelines) and [capabilities](https://learn.microsoft.com/en-us/azure/devops/pipelines/agents/agents?view=azure-devops&tabs=yaml%2Cbrowser#capabilities).

So far, the PoC is working in environments with 2 to 4 tenants. However, the project is not yet complete and will continue to develop over the coming months.

**Comment by raandree** [comment]
Yes, one build agent per tenant (self hosted as Microsoft hosted ones do not support managed identities) each assigned a managed identity with the necessary permissions in the tenant.

**Comment by andikrueger** [comment]
@Mohankade Does this solve your issue? Do you need any further support?

**Comment by Mohankade** [comment]
Thank You @andikrueger and @raandree for your support. in our case, we are using azure automation account with hybrid worker. So I am going to test this with Automation account managed identity. I will keep posted with my updates here.

**Comment by Mohankade** [comment]
Hello @andikrueger and @raandree. I have been testing the M365DSC from Azure Automation account with Hybrid worker. I have done the following configurations:

Created the Hybrid Worker VM
Created The Azure Automation Account
Created Hybrid worker group
enabled system identity for automation account and granted the grap permissions to run the DSC workloads
Created a Runbook with DSC configuration file.

The runbook is creating a mof file, however when I try to apply this configuration using start-dscconfiguration command. it fails.  I am using managed service Identity.

Configuration AADAuthorizationPolicy
{
    param (
    )

 
    Import-DscResource -ModuleName 'Microsoft365DSC' -ModuleVersion '1.23.1122.1'

    Node localhost
    {
        AADAuthorizationPolicy "AADAuthorizationPolicy"
        {
            AllowedToSignUpEmailBasedSubscriptions                  = $True;
            AllowedToUseSSPR                                        = $True;
            AllowEmailVerifiedUsersToJoinOrganization               = $True;
            AllowInvitesFrom                                        = "adminsAndGuestInviters";
            ManagedIdentity                                         = $True
            BlockMsolPowerShell                                     = $False;
            DefaultUserRoleAllowedToCreateApps                      = $False;
            DefaultUserRoleAllowedToCreateSecurityGroups            = $False;
            DefaultUserRoleAllowedToCreateTenants                   = $False;
            DefaultUserRoleAllowedToReadBitlockerKeysForOwnedDevice = $True;
            DefaultUserRoleAllowedToReadOtherUsers                  = $True;
            Description                                             = "Used to manage authorization related settings across the company.";
            DisplayName                                             = "Authorization Policy";
            Ensure                                                  = "Present";
            GuestUserRole                                           = "RestrictedGuest";
            IsSingleInstance                                        = "Yes";
            PermissionGrantPolicyIdsAssignedToDefaultUserRole       = @();
            TenantId                                                = "mytenantname";
        }
    }
}

#AADAuthorizationPolicy
AADAuthorizationPolicy  

$Path=Get-Location
$ConfigurationPath="$Path\AADAuthorizationPolicy\localhost.mof"

$ConfigurationPath

Copy-Item $ConfigurationPath -Destination "C:\Temp\AADAuthorizationPolicy"

$job=Start-DscConfiguration -Path 'C:\Temp\AADAuthorizationPolicy'

do{
Write-Output "The Tenant Configuration is being Applied"

}while($job.State -eq 'running')

$job.State

$job

$job.JobStateInfo
$job.Error


I am doing anything wrong.

**Comment by andikrueger** [comment]
Could you share the detailed error message?

**Comment by Mohankade** [comment]
The runbook is returning following error.

Directory: C:\ProgramData\Microsoft\System 
    Center\Orchestrator\7.2\SMA\Sandboxes\30a2bv5p.tiy\Temp\g0ls2dju.dkx\AADAuthorizationPolicy
Mode                LastWriteTime         Length Name                                                                   
----                -------------         ------ ----                                                                   
-a----       11/27/2023   4:56 PM           3556 localhost.mof                                                          

C:\ProgramData\Microsoft\System Center\Orchestrator\7.2\SMA\Sandboxes\30a2bv5p.tiy\Temp\g0ls2dju.dkx\AADAuthorizationPolicy\localhost.mof

The Tenant Configuration is being Applied

The Tenant Configuration is being Applied

The Tenant Configuration is being Applied

The Tenant Configuration is being Applied

The Tenant Configuration is being Applied

The Tenant Configuration is being Applied

The Tenant Configuration is being Applied

The Tenant Configuration is being Applied

The Tenant Configuration is being Applied

The Tenant Configuration is being Applied

The Tenant Configuration is being Applied

The Tenant Configuration is being Applied

The Tenant Configuration is being Applied

Failed


Id     Name            PSJobTypeName   State         HasMoreData     Location             Command                  
--     ----            -------------   -----         -----------     --------             -------                  
1      Job1            Configuratio... Failed        True            localhost            $job=Start-DscConfigur...


 State Reason
 ----- ------
Failed

**Comment by Mohankade** [comment]
The issue was with the Microsoft365DSC modules installed on Automation Account. I have re-installed the modules and issue was fixed.
